algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs

```diff
@@ -12,58 +12,268 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use core::convert::TryInto;
 use std::collections::BTreeMap;
 
 use crate::{
-    fft::DensePolynomial,
+    fft::{
+        domain::{FFTPrecomputation, IFFTPrecomputation},
+        polynomial::PolyMultiplier,
+        DensePolynomial,
+        EvaluationDomain,
+        Evaluations as EvaluationsOnDomain,
+    },
     polycommit::sonic_pc::{LabeledPolynomial, PolynomialInfo, PolynomialLabel},
     snark::marlin::{
-        ahp::{verifier, AHPError, AHPForR1CS},
+        ahp::{indexer::CircuitInfo, verifier, AHPError, AHPForR1CS, CircuitId},
+        matrices::MatrixArithmetization,
         prover,
+        witness_label,
         MarlinMode,
     },
 };
+use snarkvm_fields::{batch_inversion_and_mul, PrimeField};
+use snarkvm_utilities::{cfg_iter, cfg_iter_mut, ExecutionPool};
 
 use itertools::Itertools;
 use rand_core::RngCore;
+
+#[cfg(not(feature = "serial"))]
 use rayon::prelude::*;
-use snarkvm_fields::PrimeField;
+
+type Sum<F> = F;
+type Lhs<F> = DensePolynomial<F>;
+type Gpoly<F> = LabeledPolynomial<F>;
 
 impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-    /// Output the number of oracles sent by the prover in the third round.
-    pub fn num_fourth_round_oracles() -> usize {
-        1
+    /// Output the number of oracles sent by the prover in the fourth round.
+    pub fn num_fourth_round_oracles(circuits: usize) -> usize {
```

Comment:
```suggestion
    pub const fn num_fourth_round_oracles(circuits: usize) -> usize {
```

Reply:
I assume this comment was made in a past version when we didn't pass arguments to the function yet.

Reply:
indeed, but this one should still be `const`-able, unless the other implementations are more advanced